File: backend/src/models/organization.ts
import {
  BaseEntity,
  Column,
  CreateDateColumn,
  Entity,
  Index,
  ManyToMany,
  ManyToOne,
  OneToMany,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
  Relation
} from 'typeorm';
import { Domain, OrganizationTag, Role, Scan, ScanTask, User } from './index';

export interface PendingDomain {
  name: string;
  token: string;
}

@Entity()
export class Organization extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  @Index({ unique: true })
  @Column({
    nullable: true,
    unique: true
  })
  acronym: string;

  @Column()
  name: string;

  @Column('varchar', { array: true })
  rootDomains: string[];

  @Column('varchar', { array: true })
  ipBlocks: string[];

  @Column()
  isPassive: boolean;

  @OneToMany((type) => Domain, (domain) => domain.organization, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  domains: Relation<Domain>[];

  @Column('jsonb', { default: '[]' })
  pendingDomains: Relation<PendingDomain>[];

  @OneToMany((type) => Role, (role) => role.organization, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  userRoles: Relation<Role>[];

  /**
   * Corresponds to "organization" property of ScanTask.
   * Deprecated, replaced by "allScanTasks" property.
   */
  @OneToMany((type) => ScanTask, (scanTask) => scanTask.organization, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  scanTasks: Relation<ScanTask>[];

  /**
   * Corresponds to "organizations" property of ScanTask.
   */
  @ManyToMany((type) => ScanTask, (scanTask) => scanTask.organizations, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  allScanTasks: Relation<ScanTask>[];

  @ManyToMany((type) => Scan, (scan) => scan.organizations, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  granularScans: Relation<Scan>[];

  @ManyToMany((type) => OrganizationTag, (tag) => tag.organizations, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  tags: Relation<OrganizationTag>[];

  /**
   * The organization's parent organization, if any.
   * Organizations without a parent organization are
   * shown to users as 'Organizations', while organizations
   * with a parent organization are shown to users as 'Teams'
   */
  @ManyToOne((type) => Organization, (org) => org.children, {
    onDelete: 'CASCADE',
    nullable: true
  })
  parent: Relation<Organization>;

  @OneToMany((type) => Organization, (org) => org.parent, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  children: Relation<Organization>[];

  @ManyToOne((type) => User, {
    onDelete: 'SET NULL',
    onUpdate: 'CASCADE'
  })
  createdBy: Relation<User>;

  @Column({
    nullable: true
  })
  country: string;

  @Column({
    nullable: true
  })
  state: string;

  @Column({
    nullable: true
  })
  regionId: string;

  @Column({
    nullable: true
  })
  stateFips: number;

  @Column({
    nullable: true
  })
  stateName: string;

  @Column({
    nullable: true
  })
  county: string;

  @Column({
    nullable: true
  })
  countyFips: number;

  @Column({
    nullable: true
  })
  type: string;
}

File: backend/src/xfd_django/xfd_api/schema_models/organization.py
"""Schemas to support Organization endpoints."""

# Standard Python Libraries
from datetime import datetime
from typing import Any, List, Optional
from uuid import UUID

# Third-Party Libraries
from pydantic import BaseModel

from .organization_tag import OrganizationalTags


class Organization(BaseModel):
    """Organization schema reflecting model."""

    id: UUID
    createdAt: datetime
    updatedAt: datetime
    acronym: Optional[str]
    name: str
    rootDomains: List[str]
    ipBlocks: List[str]
    isPassive: bool
    pendingDomains: Optional[List[dict]]
    country: Optional[str]
    state: Optional[str]
    regionId: Optional[str]
    stateFips: Optional[int]
    stateName: Optional[str]
    county: Optional[str]
    countyFips: Optional[int]
    type: Optional[str]


class UserRoleSchema(BaseModel):
    """User role schema."""

    id: UUID
    role: str
    approved: bool


class TagSchema(BaseModel):
    """Tag schema."""

    id: UUID
    createdAt: datetime
    updatedAt: datetime
    name: str


class GetTagSchema(BaseModel):
    """Tag simplified schema."""

    id: UUID
    name: str


class SimpleScanSchema(BaseModel):
    """Simple scan schema."""

    id: UUID
    name: str


class GranularScanSchema(BaseModel):
    """Granular task schema."""

    id: UUID
    createdAt: datetime
    updatedAt: datetime
    name: str
    arguments: Any
    frequency: int
    lastRun: Optional[datetime]
    isGranular: bool
    isUserModifiable: Optional[bool]
    isSingleScan: bool
    manualRunPending: bool
    tags: Optional[List[OrganizationalTags]] = []
    organizations: Optional[List[Organization]] = []


class ScanTaskSchema(BaseModel):
    """Scan task schema."""

    id: UUID
    createdAt: datetime
    scan: SimpleScanSchema


class GetOrganizationSchema(BaseModel):
    """Schema for listing an organization."""

    id: UUID
    createdAt: datetime
    updatedAt: datetime
    acronym: Optional[str]
    name: str
    rootDomains: List[str]
    ipBlocks: List[str]
    isPassive: bool
    pendingDomains: Optional[Any] = []
    country: Optional[str] = None
    state: Optional[str] = None
    regionId: Optional[str] = None
    stateFips: Optional[int] = None
    stateName: Optional[str] = None
    county: Optional[str] = None
    countyFips: Optional[int] = None
    type: Optional[str] = None
    userRoles: Optional[List[UserRoleSchema]] = []
    tags: Optional[List[TagSchema]] = []


class GetSingleOrganizationSchema(BaseModel):
    """Schema for listing an organization."""

    id: UUID
    createdAt: datetime
    updatedAt: datetime
    acronym: Optional[str]
    name: str
    rootDomains: List[str]
    ipBlocks: List[str]
    isPassive: bool
    createdBy: Optional[Any] = {}
    pendingDomains: Optional[Any] = []
    country: Optional[str] = None
    state: Optional[str] = None
    regionId: Optional[str] = None
    stateFips: Optional[int] = None
    stateName: Optional[str] = None
    county: Optional[str] = None
    countyFips: Optional[int] = None
    type: Optional[str] = None
    userRoles: Optional[List[UserRoleSchema]] = []
    tags: Optional[List[TagSchema]] = []
    parent: Optional[Any] = {}
    children: Optional[Any] = {}
    granularScans: Optional[List[GranularScanSchema]] = []
    scanTasks: Optional[List[ScanTaskSchema]] = []


class NewTag(BaseModel):
    """Schema for tag data."""

    name: str  # Adjust this if there could be an 'id' field


class NewOrganization(BaseModel):
    """Create a new organization schema."""

    acronym: Optional[str]
    name: str
    rootDomains: List[str]
    ipBlocks: List[str]
    isPassive: bool
    pendingDomains: Optional[Any] = []
    country: Optional[str] = None
    state: Optional[str] = None
    regionId: Optional[str] = None
    stateFips: Optional[int] = None
    stateName: Optional[str] = None
    county: Optional[str] = None
    countyFips: Optional[int] = None
    type: Optional[str] = None
    parent: Optional[str] = None
    tags: Optional[List[NewTag]] = None


class NewOrgUser(BaseModel):
    """Add a user to organization schema."""

    userId: str
    role: str


class NewOrgScan(BaseModel):
    """Update an organization scan schema."""

    enabled: bool


class RegionSchema(BaseModel):
    """Update an organization scan schema."""

    regionId: str


class GenericMessageResponseModel(BaseModel):
    """Generic response model."""

    status: str
    message: str


class GenericPostResponseModel(BaseModel):
    """Generic response model."""

    statusCode: int
    body: Any
